contracts/src/mcms.cairo

```diff
@@ -0,0 +1,610 @@
+                'wrong chain id'
+            );
+            assert(
+                starknet::info::get_contract_address() == metadata.multisig,
+                'wrong multisig address'
+            );
+
+            let op_count = self.s_expiring_root_and_op_count.read().op_count;
+            let current_root_metadata = self.s_root_metadata.read();
+            assert(
+                op_count == current_root_metadata.post_op_count
+                    || current_root_metadata.override_previous_root,
+                'expect pending operations'
+            );
+            assert(op_count == metadata.pre_op_count, 'wrong pre-operation count');
+            assert(metadata.pre_op_count <= metadata.post_op_count, 'wrong post-operation count');
+
+            self.s_seen_signed_hashes.write(msg_hash, true);
+            self
+                .s_expiring_root_and_op_count
+                .write(
+                    ExpiringRootAndOpCount {
+                        root: root, valid_until: valid_until, op_count: metadata.pre_op_count
+                    }
+                );
+
+            self
+                .emit(
+                    Event::NewRoot(
+                        NewRoot { root: root, valid_until: valid_until, metadata: metadata, }
+                    )
+                );
+        }
+
+        fn execute(ref self: ContractState, op: Op, proof: Span<u256>) {
+            let current_expiring_root_and_op_count = self.s_expiring_root_and_op_count.read();
+
+            assert(
+                self
+                    .s_root_metadata
+                    .read()
+                    .post_op_count > current_expiring_root_and_op_count
+                    .op_count,
+                'post-operation count reached'
+            );
+
+            assert(
+                starknet::info::get_tx_info().unbox().chain_id.into() == op.chain_id,
+                'wrong chain id'
+            );
+
+            assert(starknet::info::get_contract_address() == op.multisig, 'wrong multisig address');
+
+            assert(
+                current_expiring_root_and_op_count
+                    .valid_until
+                    .into() >= starknet::info::get_block_timestamp(),
+                'root has expired'
+            );
+
+            assert(op.nonce == current_expiring_root_and_op_count.op_count, 'wrong nonce');
+
+            // verify op exists in merkle tree
+            let mut encoded_leaf: Bytes = BytesTrait::new_empty()
+                .encode(MANY_CHAIN_MULTI_SIG_DOMAIN_SEPARATOR_OP)
+                .encode(op.chain_id)
+                .encode(op.multisig)
+                .encode(op.nonce)
+                .encode(op.to)
+                .encode(op.selector);
+            // encode the data field by looping through
+            let mut i = 0;
+            while i < op.data.len() {
+                encoded_leaf = encoded_leaf.encode(*op.data.at(i));
```

Comment:
just checking if this encoding does any padding on each element of data? felt252 is smaller than u256 so might result in a different encoding if so

Reply:
all of the `.encode()` do no padding similar to `abi.encode` for solidity, so we'll fit the current standards

(there is an `.encode_packed` if you'd like to do `abi.encodePacked`)

Reply:
I think `abi.encode` does do padding - [inputs less than 32 bytes are padded to 32 bytes](https://medium.com/@scourgedev/deep-dive-into-abi-encode-types-padding-and-disassembly-84472f1b4543#5853). `abi.encodePacked` is the non-padding encoding

Reply:
sorry yeah I meant *.encode does padding

Reply:
yep so i was wondering if the data will be encoded here like `encoded = (0x00.. + felt252(chunk1)) + (0x00.. + felt252(chunk2)) + ...`. We might want to be using encode_packed instead, since data should already be encoded calldata?

Reply:
The calldata is actually just an array of felts, so either encode or encode_packed should work. 

Since every other encoding is unpacked (on Starknet and also on EVM) I don't see a reason to use packed encoding specifically for this case.

Reply:
Ah ok i think i understand now - i was confused as to how calldata is handled on Starknet.
Calldata on EVM and most other chains usually represents the encoded raw data to be passed in a transaction and is in the format of bytes or a hex string, e.g. EVM calldata is `0x + <4 bytes func selector> + <encoded args>`. However on Starknet it seems like calldata is more akin to "raw args" rather than encoded data and is represented by an array of felts corresponding to args (or arg metadata). I realized that since there's no encoding necessary as it gets passed straight to the called contract, you're right - we are free to choose the encoding method here for the purposes of generating a hash for MCM and just have to make sure it is consistent between the contract and tooling.